File: python/004.py

def nums(mul=2):
    start = pow(10, mul) - 1
    stop = start - pow(10, mul - 1)
    rng = xrange(start, stop, -1)
    for i in rng:
        for j in rng:
            t = "%d" % (i * j)
            if t == t[::-1]:
                yield i * j

if __name__ == '__main__':
    assert max(nums(2)) == 9009

    assert max(nums(3)) == 906609

File: ruby/004.rb
def solve(mul=2)
    max = 0

    rng = (1...10**mul).to_a
    rng.product(rng) do |i, j|
       v = i*j
       if "#{i*j}".reverse.to_i == v
           if v > max
               max = v
           end
       end
    end
    max
end


require 'test/unit'
class TestSolve < Test::Unit::TestCase
    def test_nums
        assert solve(2) == 9009
        assert solve(3) == 906609
    end
end
